libnymea-core/scriptengine/scriptengine.cpp

```diff
@@ -0,0 +1,466 @@
+#include "scriptalarm.h"
+
+#include "nymeasettings.h"
+
+#include <QQmlApplicationEngine>
+#include <QQmlContext>
+#include <QQmlComponent>
+#include <QJsonParseError>
+#include <QJsonDocument>
+
+#include "loggingcategories.h"
+
+#include <QDir>
+
+namespace nymeaserver {
+
+QList<ScriptEngine*> ScriptEngine::s_engines;
+QtMessageHandler ScriptEngine::s_upstreamMessageHandler;
+QLoggingCategory::CategoryFilter ScriptEngine::s_oldCategoryFilter = nullptr;
+
+ScriptEngine::ScriptEngine(DeviceManager *deviceManager, QObject *parent) : QObject(parent),
+    m_deviceManager(deviceManager)
+{
+    qmlRegisterType<ScriptEvent>("nymea", 1, 0, "DeviceEvent");
+    qmlRegisterType<ScriptAction>("nymea", 1, 0, "DeviceAction");
+    qmlRegisterType<ScriptState>("nymea", 1, 0, "DeviceState");
+    qmlRegisterType<ScriptAlarm>("nymea", 1, 0, "Alarm");
+
+    m_engine = new QQmlEngine(this);
+    m_engine->setProperty("deviceManager", reinterpret_cast<quint64>(m_deviceManager));
+
+    // Don't automatically print script warnings (that is, runtime errors, *not* console.warn() messages)
+    // to stdout as they'd end up on the "default" logging category.
+    // We collect them ourselves through the warnings() signal and print them to the dcScriptEngine category.
+    m_engine->setOutputWarningsToStandardError(false);
+    connect(m_engine, &QQmlEngine::warnings, this, [this](const QList<QQmlError> &warnings){
+        foreach (const QQmlError &warning, warnings) {
+            QMessageLogContext ctx(warning.url().toString().toUtf8(), warning.line(), "", "ScriptEngine");
+            // Send to script logs
+            onScriptMessage(
+#if QT_VERSION >= QT_VERSION_CHECK(5,9,0)
+                        warning.messageType(),
+#else
+                        QtMsgType::QtWarningMsg,
+#endif
+                        ctx, warning.description());
+            // and to logging system
+            qCWarning(dcScriptEngine()) << warning.toString();
+        }
+    });
+
+    // console.log()/warn() messages instead are printed to the "qml" category. We install our own
+    // filter to *always* get them, regardless of the configured logging categories
+    if (!s_oldCategoryFilter) {
+        s_oldCategoryFilter = QLoggingCategory::installFilter(&logCategoryFilter);
+    }
+    // and our own handler to redirect them to the ScriptEngine category
+    if (s_engines.isEmpty()) {
+        s_upstreamMessageHandler = qInstallMessageHandler(&logMessageHandler);
+    }
+    s_engines.append(this);
+
+
+    QDir dir;
+    if (!dir.exists(NymeaSettings::storagePath() + "/scripts/")) {
+        dir.mkpath(NymeaSettings::storagePath() + "/scripts/");
+    }
+
+    loadScripts();
+
+}
+
+ScriptEngine::~ScriptEngine()
+{
+    s_engines.removeAll(this);
+    if (s_engines.isEmpty()) {
+        qInstallMessageHandler(s_upstreamMessageHandler);
+    }
+}
+
+Scripts ScriptEngine::scripts()
+{
+    Scripts ret;
+    foreach (Script *script, m_scripts) {
+        ret.append(*script);
```

Comment:
`(*script)` ?

Reply:
well, m_scripts is a list of `Script*` but I won't pass the internal list of objects to the outside of the class so I copy the entries to a by-value list.